File: member-service/src/main/java/com/hedvig/memberservice/web/AuthController.java
package com.hedvig.memberservice.web;

import com.hedvig.external.bankID.bankidTypes.CollectResponse;
import com.hedvig.external.bankID.bankidTypes.OrderResponse;
import com.hedvig.external.bankID.bankidTypes.ProgressStatus;
import com.hedvig.memberservice.aggregates.exceptions.BankIdReferenceUsedException;
import com.hedvig.memberservice.commands.AuthenticationAttemptCommand;
import com.hedvig.memberservice.commands.BankIdAuthenticationStatus;
import com.hedvig.memberservice.commands.BankIdSignCommand;
import com.hedvig.memberservice.commands.InactivateMemberCommand;
import com.hedvig.memberservice.query.*;
import com.hedvig.memberservice.services.BankIdService;
import com.hedvig.memberservice.web.dto.*;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.annotation.*;

import java.io.UnsupportedEncodingException;
import java.util.Objects;
import java.util.Optional;

import static net.logstash.logback.argument.StructuredArguments.value;

@RestController()
@RequestMapping("/member/bankid/")
public class AuthController {

    private final CommandGateway commandGateway;
    private final MemberRepository memberRepo;
    private final SignedMemberRepository signedMemberRepository;
    private final CollectRepository collectRepo;
    private final BankIdService bankIdService;
    private static Logger log = LoggerFactory.getLogger(AuthController.class);

    @Autowired
    public AuthController(CommandGateway commandGateway,
                          MemberRepository memberRepo,
                          SignedMemberRepository signedMemberRepository,
                          CollectRepository collectRepo,
                          BankIdService bankIdService) {
        this.commandGateway = commandGateway;
        this.memberRepo = memberRepo;
        this.signedMemberRepository = signedMemberRepository;
        this.collectRepo = collectRepo;
        this.bankIdService = bankIdService;
    }

    @PostMapping(path = "auth")
    public ResponseEntity<BankIdAuthResponse> auth(@RequestBody BankIdAuthRequest request) {

        log.info("Auth request for with: {}", request.getSsn(), value("memberId", request.getMemberId()));

        long memberId = convertMemberId(request.getMemberId());

        OrderResponse status = bankIdService.auth(request.getSsn(), memberId);
        BankIdAuthResponse response = new BankIdAuthResponse(status.getAutoStartToken(), status.getOrderRef());


        return ResponseEntity.ok(response);
    }

    @PostMapping(path ="sign")
    public ResponseEntity<BankIdSignResponse> sign(@RequestBody BankIdSignRequest request) throws UnsupportedEncodingException {
        long memberId = convertMemberId(request.getMemberId());

        log.info("Sign request for ssn: {}", request.getSsn(), value("memberId", request.getMemberId()));

        OrderResponse status = bankIdService.sign(request.getSsn(), request.getUserMessage(), memberId);
        BankIdSignResponse response = new BankIdSignResponse(status.getAutoStartToken(), status.getOrderRef());

        return ResponseEntity.ok(response);
    }

    private long convertMemberId(String memberId) {
        try {
            return Long.parseLong(memberId);
        }catch(Exception e) {
            throw new HttpMessageNotReadableException("Could not parse memberId");
        }
    }



    @PostMapping(path = "collect")
    public ResponseEntity<?> collect(@RequestParam String referenceToken, @RequestHeader(value = "hedvig.token") Long hid) throws InterruptedException {

        log.info("Start collect");

        CollectType collectType = collectRepo.findOne(referenceToken);
        BankIdCollectResponse response;
        
        if(collectType==null){
        	log.error("ERROR: Oh no! Collect type is null!");
        	return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
        
        if(collectType.type.equals(CollectType.RequestType.AUTH)) {
            CollectResponse status = bankIdService.authCollect(referenceToken);


            if (status.getProgressStatus() == ProgressStatus.COMPLETE) {
                String ssn = status.getUserInfo().getPersonalNumber();

                Optional<SignedMemberEntity> member = signedMemberRepository.findBySsn(ssn);

                Long currentMemberId = hid;
                if (member.isPresent()) {
                    SignedMemberEntity m = member.get();
                    if (!m.getId().equals(hid)) {
                        this.commandGateway.sendAndWait(new InactivateMemberCommand(hid));
                    }
                    currentMemberId = m.getId();
                }

                try {
                    BankIdAuthenticationStatus authStatus = new BankIdAuthenticationStatus();
                    authStatus.setSSN(status.getUserInfo().getPersonalNumber());
                    authStatus.setGivenName(status.getUserInfo().getGivenName());
                    authStatus.setSurname(status.getUserInfo().getSurname());
                    authStatus.setReferenceToken(referenceToken);
                    this.commandGateway.sendAndWait(new AuthenticationAttemptCommand(currentMemberId, authStatus));
                    Thread.sleep(1000L);
                } catch (BankIdReferenceUsedException e) {
                    log.info("Old reference token used: ", e);
                    return ResponseEntity.badRequest().body("{\"message\":\"" + e.getMessage() + "\"}");
                }

                response = new BankIdCollectResponse(BankIdProgressStatus.valueOf(status.getProgressStatus().name()), referenceToken, Objects.toString(currentMemberId));

                return ResponseEntity.ok().header("Hedvig.Id", currentMemberId.toString()).body(response);
            }

            return ResponseEntity.ok(new BankIdCollectResponse(BankIdProgressStatus.valueOf(status.getProgressStatus().name()),  referenceToken, hid.toString()));

        } else if (collectType.type.equals(CollectType.RequestType.SIGN)) {
            CollectResponse status = bankIdService.signCollect(referenceToken);
            if(status.getProgressStatus() == ProgressStatus.COMPLETE) {
                Optional<MemberEntity> memberEntity = memberRepo.findById(hid);
                if (memberEntity.isPresent()) {
                    this.commandGateway.sendAndWait(new BankIdSignCommand(hid, referenceToken, status.getSignature(), status.getOcspResponse()));
                }
            }

            return ResponseEntity.ok(new BankIdCollectResponse(BankIdProgressStatus.valueOf(status.getProgressStatus().name()),  referenceToken, hid.toString()));
        } else {
            return ResponseEntity.noContent().build();
        }
    }
}

File: member-service/src/main/java/com/hedvig/memberservice/web/AuthController.kt
package com.hedvig.memberservice.web


import com.hedvig.external.authentication.dto.StartZignSecAuthenticationResult
import com.hedvig.memberservice.commands.models.ZignSecAuthenticationMarket
import com.hedvig.memberservice.services.BankIdService
import com.hedvig.memberservice.services.signing.zignsec.ZignSecBankIdService
import com.hedvig.memberservice.util.getEndUserIp
import com.hedvig.memberservice.web.dto.BankIdAuthCountry
import com.hedvig.memberservice.web.dto.BankIdAuthRequest
import com.hedvig.memberservice.web.dto.BankIdAuthResponse
import com.hedvig.memberservice.web.dto.GenericBankIdAuthenticationRequest
import com.hedvig.memberservice.web.dto.ZignSecStartDto
import net.logstash.logback.argument.StructuredArguments
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/member/")
class AuthController @Autowired constructor(
    private val bankIdService: BankIdService,
    private val zignSecBankIdService: ZignSecBankIdService
) {

    @PostMapping(path = ["bankid/auth"])
    fun auth(@RequestHeader(value = "x-forwarded-for", required = false) forwardedIp: String?, @RequestBody request: BankIdAuthRequest): ResponseEntity<BankIdAuthResponse> {
        MDC.put("memberId", request.memberId)
        log.info(
            "Auth request for with memberId: ${request.memberId}", StructuredArguments.value("memberId", request.memberId))

        val memberId = convertMemberId(request.memberId)

        val endUserIp = forwardedIp
            .getEndUserIp("Header 'x-forwarded-for' was not included when calling AuthController auth! MemberId:$memberId")

        val status = bankIdService.auth(memberId, endUserIp)

        val response = BankIdAuthResponse(
            autoStartToken = status.autoStartToken,
            referenceToken = status.orderRef
        )

        return ResponseEntity.ok(response)
    }

    private fun convertMemberId(memberId: String): Long {
        return try {
            memberId.toLong()
        } catch (e: Exception) {
            throw HttpMessageNotReadableException("Could not parse memberId")
        }
    }

    @PostMapping(path = ["/{country}/bankid/auth"])
    private fun auth(
        @RequestHeader("hedvig.token") memberId: Long,
        @RequestHeader(value = "Accept-Language", required = false) acceptLanguage: String?,
        @RequestHeader("Authorization") authorization: String,
        @PathVariable("country") country: BankIdAuthCountry,
        @RequestBody request: GenericBankIdAuthenticationRequest
    ): ResponseEntity<ZignSecStartDto> {
        MDC.put("memberId", memberId.toString())

        val result = zignSecBankIdService.authenticate(
            memberId = memberId,
            personalNumber = request.personalNumber,
            market = when (country) {
                BankIdAuthCountry.norway -> ZignSecAuthenticationMarket.NORWAY
                BankIdAuthCountry.denmark -> ZignSecAuthenticationMarket.DENMARK
            },
            acceptLanguage = acceptLanguage
        )

        return when (result) {
            is StartZignSecAuthenticationResult.Success ->
                ResponseEntity.ok(ZignSecStartDto(result.redirectUrl))
            is StartZignSecAuthenticationResult.StaticRedirect ->
                ResponseEntity.ok(ZignSecStartDto(result.redirectUrl))
            is StartZignSecAuthenticationResult.Failed -> {
                log.warn("ZignSec authentication start failed, codes = ${result.errors.map { it.code }}")
                ResponseEntity.status(HttpStatus.FORBIDDEN).build()
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(AuthController::class.java)
    }

}

